### PR TITLE
Feature/10 clean session store files

### PR DESCRIPTION
**Pull request by ppxl**
solves #10 and #11 

This PR introduces a background cleaning task for files created by the `FileSessionStore`.
- The task iterates all JWTs, and filters for the expiration date. Then it proceeds to remove the JWTs and the associated service ticket file
- The task run interval is configurable with the property `sonar.cas.sessionStore.cleanUpIntervalInSeconds`.
- Especially for the file removal great lengths were taken (unit test-wise) to ensure only the right files are removed.

This PR adds additionally more unit and integration tests, and elaborates the documentation quite a bit

As usual, larger code parts were extracted and isolated for better maintainability and readability. Most specifically:
- CasIdentityProvider functionality is split into LoginHandler and LogoutHandler
- TicketValidation parts extract into a TicketValidationFactory

### CODE REVIEW

File: architecture.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+- an expiration date
+
+### Service Ticket
+
+The service ticket is a code which is used to validate the user's authentication request directly with the 
+authenticating system, namely the CAS.
+
+## Use cases within the authentication lifecycle
+
+lorem ipsum
```
Comment (sdorra) [inline architecture.md]: lorem ipsum?

File: pluginConfiguration.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+### Specifies whether gateway=true should be sent to the CAS server.
+
+Default is false.
+
+`sonar.cas.sendGateway=false`
+
+### Path to CAS Session Store
+
+As SonarQube does not provide a session (other than by issueing JWT tokens). When the user
+logs out, the cookie containing the necessary JWT token is removed. WhatBut SonarQube *DOES
```
Comment (sdorra) [inline pluginConfiguration.md]: WhatBut?

File: pluginConfiguration.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+interval in seconds between each clean up run. Do not set the interval too short (this could lead to unnecessary
+CPU load) or too long (this could lead to unnecessary filesystem load).
+  
+Default is 30 minutes, 0 disables the cleanup (this SHOULD NOT be done in a production environment)
+
+`sonar.cas.sessionStore.cleanUpIntervalInSeconds = 1800`
+
+### Configure CAS Attribute(s) 
+
+Attributes holding the authorities (groups, roles, etc.) the user belongs to. Multiple
```
Comment (sdorra) [inline pluginConfiguration.md:82]: Looks like a missing headline for the role attribute

File: pluginConfiguration.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+`sonar.cas.saml11.toleranceMilliseconds=1000`
+
+### Ignore certification validation errors.
+
+**CAUTION! NEVER USE IN PROD! SECURITY RISK!**
+
+This is only for development environments where a proper certificate chain is unfeasible. 
+
+`sonar.cas.disableCertValidation=false`
```
Comment (sdorra) [inline pluginConfiguration.md]: Description for `urlAfterCasRedirectCookieMaxAgeSeconds` is missing

File: src/main/java/org/sonar/plugins/cas/CasIdentityProvider.java [hunk 5]
```diff
@@ -84,109 +72,34 @@ public void init(Context context) {
             HttpServletRequest request = context.getRequest();
 
             if (isLogin(request)) {
-                handleAuthentication(context);
+                LOG.debug("Found internal login case");
+                new LoginHandler(configuration, attributes, casSessionStore, validatorFactory).handleLogin(context);
```
Comment (sdorra) [inline src/main/java/org/sonar/plugins/cas/CasIdentityProvider.java]: I think we could create an extension for the `LoginHandler` and `LogoutHandler` which are getting the required dependencies via injection. Doing so we could move most of the constructor injections of `CasIdentityProvider` into the `LoginHandler` and  `LogoutHandler`.

File: src/main/java/org/sonar/plugins/cas/CasIdentityProvider.java [hunk 6]
```diff
@@ -84,109 +72,34 @@ public void init(Context context) {
...
             } else if (isLogout(request)) {
-                handleLogout(context);
+                LOG.debug("Found external logout case");
+                new LogoutHandler(configuration, casSessionStore).logout(context.getRequest(), context.getResponse());
```
Comment (sdorra) [inline src/main/java/org/sonar/plugins/cas/CasIdentityProvider.java]: See comment above.

File: src/main/java/org/sonar/plugins/cas/CasPlugin.java [hunk 7]
```diff
@@ -55,18 +53,14 @@ public void define(Context context) {
...
-            String protocol = "cas2";
-
-            Preconditions.checkState(!Strings.isNullOrEmpty(protocol),
-                    "Missing CAS protocol. Values are: cas1, cas2 or saml11.");
-
-            // The ignore certification validation should only be used in development (security risk)!
-            // if (configuration.mustGetBoolean(SonarCasProperties.DISABLE_CERT_VALIDATION.toString()).orElse(Boolean.FALSE)) {
-            IgnoreCert.disableSslVerification();
 
+        if (isRealmEnabled()) {
```
Comment (sdorra) [inline src/main/java/org/sonar/plugins/cas/CasPlugin.java]: i think we could remove this

File: src/main/java/org/sonar/plugins/cas/CasPlugin.java [hunk 8]
```diff
@@ -83,8 +77,6 @@ public void define(Context context) {
     }
 
     private boolean isRealmEnabled() {
```
Comment (sdorra) [inline src/main/java/org/sonar/plugins/cas/CasPlugin.java]: i think we could remove this

File: src/main/java/org/sonar/plugins/cas/CasRestClient.java [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+import java.io.*;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.net.URLEncoder;
+
+/**
+ * Client for the CAS REST API.
+ *
+ * @author Sebastian Sdorra, TRIOLOGY GmbH
+ * @see <a href="https://wiki.jasig.org/display/casum/restful+api">CAS RESTful API</a>
```
Comment (sdorra) [inline src/main/java/org/sonar/plugins/cas/CasRestClient.java]: URL is outdated. The new one is https://apereo.github.io/cas/6.0.x/protocol/REST-Protocol.html

File: src/main/java/org/sonar/plugins/cas/CasRestClient.java [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+                        "could not create service ticket, web service returned " + rc);
+            }
+
+            String content;
+            BufferedReader reader = null;
+            try {
+                reader = createReader(connection);
+                content = CharStreams.toString(reader);
+            } finally {
+                Closeables.closeQuietly(reader);
```
Comment (sdorra) [inline src/main/java/org/sonar/plugins/cas/CasRestClient.java]: should be converted to try with resources

File: src/main/java/org/sonar/plugins/cas/DevelopmentServerStartHandler.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    private Configuration configuration;
+
+    public DevelopmentServerStartHandler(Configuration configuration) {
+        this.configuration = configuration;
+    }
+
+    @Override
+    public void onServerStart(Server server) {
+        if (isSslSupportDeactivated()) {
+            LOG.error("SSL certificate check is disabled. Please ENABLE SSL disabling on a production machine for " +
```
Comment (sdorra) [inline src/main/java/org/sonar/plugins/cas/DevelopmentServerStartHandler.java]: Is this an **error**? For me it looks more like a **warning**.

File: src/main/java/org/sonar/plugins/cas/ForceCasLoginFilter.java [hunk 12]
```diff
@@ -48,74 +52,61 @@
...
-        }
-
+        HttpServletRequest request = HttpUtil.toHttp(servletRequest);
+        HttpServletResponse response = HttpUtil.toHttp(servletResponse);
         String requestedURL = request.getRequestURL().toString();
 
         if (isInWhiteList(request.getServletPath()) || isAuthenticated(request)) {
             LOG.debug("Found permitted request to {}", requestedURL);
-            boolean redirectToLoginPage = new LogoutHandler(casSessionStore).handleInvalidJwtCookie(request, response);
+            boolean redirectToLoginPage = new LogoutHandler(configuration, casSessionStore).handleInvalidJwtCookie(request, response);
```
Comment (sdorra) [inline src/main/java/org/sonar/plugins/cas/ForceCasLoginFilter.java]: We could use an extension and inject the `LogoutHandler`

File: src/main/java/org/sonar/plugins/cas/session/SessionFileRemover.java [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+                .filter(file -> JWT_FILE_PATTERN.matcher(file.getFileName().toString()).matches())
+                .collect(Collectors.toList());
+    }
+
+    List<SimpleJwt> parseJwtFiles(Collection<Path> foundFiles) {
+        List<SimpleJwt> list = new ArrayList<>(foundFiles.size());
+
+        for (Path file : foundFiles) {
+            try (BufferedReader reader = Files.newBufferedReader(file, StandardCharsets.US_ASCII)) {
+                SimpleJwt jwt = JAXB.unmarshal(reader, SimpleJwt.class);
```
Comment (sdorra) [inline src/main/java/org/sonar/plugins/cas/session/SessionFileRemover.java]: JAXB throws a `RuntimeException` if the file could not be parsed. The current implementation looks like if one file could not be parsed, the session store gets never cleaned. I think we could catch the JAXB exception too.

File: src/main/java/org/sonar/plugins/cas/session/SessionStoreCleaner.java [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+import org.sonar.api.server.ServerSide;
+import org.sonar.plugins.cas.util.SonarCasProperties;
+
+import java.util.Timer;
+import java.util.TimerTask;
+
+@ServerSide
+public final class SessionStoreCleaner implements ServerStartHandler {
+    private static final Logger LOG = LoggerFactory.getLogger(SessionStoreCleaner.class);
+    private static final int SESSION_STORE_CLEANUP_INTERVAL_IN_SECS_DEFAULT = 60 * 30;
```
Comment (sdorra) [inline src/main/java/org/sonar/plugins/cas/session/SessionStoreCleaner.java]: We could use `TimeUnit` here, it should be more readable.

Comment (sdorra) [reply src/main/java/org/sonar/plugins/cas/session/SessionStoreCleaner.java]: `TimeUnit.MINUTES.toSeconds(30l)`

File: src/main/java/org/sonar/plugins/cas/session/SessionStoreCleaner.java [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    private TimerTask createTask() {
+        return new CasSessionStoreCleanerTask();
+    }
+
+    private void runTask(TimerTask task) {
+        // The CAS session store is persistent. This means there _may be_ clean-up work to tend to after a server start
+        long firstOccurrenceImmediate = 0;
+        timer.scheduleAtFixedRate(task, firstOccurrenceImmediate, cleanUpIntervalInSeconds * MILLIS);
```
Comment (sdorra) [inline src/main/java/org/sonar/plugins/cas/session/SessionStoreCleaner.java]: We could use `TimeUnit`.

**Review by sdorra — CHANGES_REQUESTED** [review]

File: architecture.md (hunk 1 above)
Comment (ppxl) [reply architecture.md]: https://thumbs.gfycat.com/OblongTiredGonolek-size_restricted.gif

Comment (ppxl) [reply architecture.md]: sweating

Comment (ppxl) [reply architecture.md]: resolved

File: pluginConfiguration.md (hunk 2 above)
Comment (ppxl) [reply pluginConfiguration.md]: resolved

File: pluginConfiguration.md (hunk 3 above)
Comment (ppxl) [reply pluginConfiguration.md:82]: Resolved

File: src/main/java/org/sonar/plugins/cas/CasIdentityProvider.java (hunk 5 above)
Comment (ppxl) [reply src/main/java/org/sonar/plugins/cas/CasIdentityProvider.java]: resolved

File: src/main/java/org/sonar/plugins/cas/DevelopmentServerStartHandler.java (hunk 11 above)
Comment (ppxl) [reply src/main/java/org/sonar/plugins/cas/DevelopmentServerStartHandler.java]: resolved

File: src/main/java/org/sonar/plugins/cas/ForceCasLoginFilter.java (hunk 12 above)
Comment (ppxl) [reply src/main/java/org/sonar/plugins/cas/ForceCasLoginFilter.java]: resolved

File: src/main/java/org/sonar/plugins/cas/session/SessionStoreCleaner.java (hunk 14 above)
Comment (ppxl) [reply src/main/java/org/sonar/plugins/cas/session/SessionStoreCleaner.java]: resolved

File: pluginConfiguration.md (hunk 4 above)
Comment (ppxl) [reply pluginConfiguration.md]: resolved

File: src/main/java/org/sonar/plugins/cas/CasIdentityProvider.java (hunk 6 above)
Comment (ppxl) [reply src/main/java/org/sonar/plugins/cas/CasIdentityProvider.java]: resolved

File: src/main/java/org/sonar/plugins/cas/CasPlugin.java (hunk 7 above)
Comment (ppxl) [reply src/main/java/org/sonar/plugins/cas/CasPlugin.java]: resolved

File: src/main/java/org/sonar/plugins/cas/CasPlugin.java (hunk 8 above)
Comment (ppxl) [reply src/main/java/org/sonar/plugins/cas/CasPlugin.java]: resolved

File: src/main/java/org/sonar/plugins/cas/CasRestClient.java (hunk 9 above)
Comment (ppxl) [reply src/main/java/org/sonar/plugins/cas/CasRestClient.java]: resolve. Also I added similar stuff to the README

File: src/main/java/org/sonar/plugins/cas/CasRestClient.java (hunk 10 above)
Comment (ppxl) [reply src/main/java/org/sonar/plugins/cas/CasRestClient.java]: `CharStreams` is regarded as unstable. I rewrote it as try with resources and do plain reading `while`. Added integration tests as well

File: src/main/java/org/sonar/plugins/cas/session/SessionFileRemover.java (hunk 13 above)
Comment (ppxl) [reply src/main/java/org/sonar/plugins/cas/session/SessionFileRemover.java]: Catches now all exceptions and should not break any more, and produces an error log.

File: src/main/java/org/sonar/plugins/cas/session/SessionStoreCleaner.java (hunk 15 above)
Comment (ppxl) [reply src/main/java/org/sonar/plugins/cas/session/SessionStoreCleaner.java]: Resolved

**Review by sdorra — APPROVED** [review]

**Pull request merged**